### DISCUSSION TITLE
Draft shell commands (needs a better name)

### DISCUSSION THREAD

**Discussion by Taitava** [Taitava/obsidian-shellcommands#286] [Ideas]
Regarding the 'New shell command' button, I'd like to add an ability to define one or multiple shell commands that act as _default_ for new shell commands. Currently, each new shell command gets default settings that are hardcoded in the plugin. Users need to always change all settings to their liking.

I don't know a good term for this, but I'll use _draft_ for now, but that will change later. Term _template_ would describe this functionality very well, but I don't want to use it, because I want to reserve it for Markdown template files used by [Templater](https://github.com/SilentVoid13/Templater)/[Templates](https://help.obsidian.md/Plugins/Templates).

In shell command settings, there would be a checkbox labelled _Use as draft shell command_ (or something like it). Being a draft shell command means:
 - Draft shell commands won't show up in command palette and cannot be executed via events nor via Shell command URI. This is due to their nature: they may be incomplete, they are just drafts that will be completed by later, new shell commands. However, executing via settings by using the _Execute now_ icon is allowed for testing purposes.
 - If any draft shell commands exist, the _New shell command_ button will disappear.
 - For every draft shell command, there will be a button that creates a new shell command based on the properties of the draft shell command. The button's label will be the same as the draft shell command's _alias_ text, so users can clearly denote what kind of shell command is created by that button. The button will also contain an icon defined in the draft shell command's settings (if set).

## Remember from which draft shell command a shell command was created
Each shell command that is derived from a draft shell command should have a non-changeable configuration field containing the id of the draft shell command that was used to create the new shell command. This would allow some future development:
 - In main settings modal: Highlight setting icons that contain customised setting values. E.g. if output settings are changed in the shell command (compared to the draft shell commend), highlight the _output_ icon.
 - Perhaps some day: Make it possible to list changes between a shell command and a draft shell command.
    - Then it's easy to see which settings were changed in the actual shell command.
    - Or, if user changes the _draft_ shell command, it would be easy to see which actual shell commands the user needs to change, too.
    - And just generally show a list of created shell commands in the draft command's settings modal.

## Candidate terms
- _Abstract_ shell command
- Shell command _blueprint_

**Comment by Cito** [comment]
The best solution might be something like OOP style inheritance. I.e. a command can inherit undefined settings from another command, and there would be one base command with all the default settings.

**Reply by Taitava** [reply]
Thank you for your input @Cito ! 🙂 I agree. There could even be a multi-level tree of commands:
```
- Root command
  - Inheritor command 1
  - Inheritor command 2
    - Inheritor command 4
  - Inheritor command 3
    - Inheritor command 5
      - etc...
```

Inheritor command would have all their settings set to `inherit from parent` mode by default. So, if a setting is changed in a parent command, it gets changed in every child that does not define the setting explicitly.

Each command could have their executability enabled/disabled independently. So even parent commands could be executed, if it makes sense for a use case.
